poseidon2-circuit-air/src/air.rs

```diff
@@ -0,0 +1,676 @@
+use alloc::vec;
+use alloc::vec::Vec;
+use core::array;
+use core::borrow::Borrow;
+
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_circuit::tables::{Poseidon2CircuitRow, Poseidon2CircuitTrace};
+use p3_field::{PrimeCharacteristicRing, PrimeField};
+use p3_matrix::Matrix;
+use p3_matrix::dense::{RowMajorMatrix, RowMajorMatrixViewMut};
+use p3_poseidon2::GenericPoseidon2LinearLayers;
+use p3_poseidon2_air::{Poseidon2Air, RoundConstants, generate_trace_rows};
+use p3_symmetric::CryptographicPermutation;
+
+use crate::sub_builder::SubAirBuilder;
+use crate::{Poseidon2CircuitCols, num_cols};
+
+/// Extends the Poseidon2 AIR with recursion circuit-specific columns and constraints.
+/// Assumes the field size is at least 16 bits.
+///
+/// SPECIFIC ASSUMPTIONS:
+/// - Memory elements from the witness table are extension elements of degree D.
+/// - RATE and CAPACITY are the number of extension elements in the rate/capacity.
+/// - WIDTH is the number of field elements in the state, i.e., (RATE + CAPACITY) * D.
+/// - `reset` can only be set during an absorb.
+#[derive(Debug)]
+pub struct Poseidon2CircuitAir<
+    F: PrimeCharacteristicRing,
+    LinearLayers,
+    const D: usize,
+    const WIDTH: usize,
+    const WIDTH_EXT: usize,
+    const RATE_EXT: usize,
+    const CAPACITY_EXT: usize,
+    const SBOX_DEGREE: u64,
+    const SBOX_REGISTERS: usize,
+    const HALF_FULL_ROUNDS: usize,
+    const PARTIAL_ROUNDS: usize,
+> {
+    p3_poseidon2: Poseidon2Air<
+        F,
+        LinearLayers,
+        WIDTH,
+        SBOX_DEGREE,
+        SBOX_REGISTERS,
+        HALF_FULL_ROUNDS,
+        PARTIAL_ROUNDS,
+    >,
+}
+
+impl<
+    F: PrimeField,
+    LinearLayers: GenericPoseidon2LinearLayers<WIDTH>,
+    const D: usize,
+    const WIDTH: usize,
+    const WIDTH_EXT: usize,
+    const RATE_EXT: usize,
+    const CAPACITY_EXT: usize,
+    const SBOX_DEGREE: u64,
+    const SBOX_REGISTERS: usize,
+    const HALF_FULL_ROUNDS: usize,
+    const PARTIAL_ROUNDS: usize,
+>
+    Poseidon2CircuitAir<
+        F,
+        LinearLayers,
+        D,
+        WIDTH,
+        WIDTH_EXT,
+        RATE_EXT,
+        CAPACITY_EXT,
+        SBOX_DEGREE,
+        SBOX_REGISTERS,
+        HALF_FULL_ROUNDS,
+        PARTIAL_ROUNDS,
+    >
+{
+    pub const fn new(
+        constants: RoundConstants<F, WIDTH, HALF_FULL_ROUNDS, PARTIAL_ROUNDS>,
+    ) -> Self {
+        assert!(CAPACITY_EXT + RATE_EXT == WIDTH_EXT);
```

Comment:
You can also do
```rust
const _: () = {
            assert!(CAPACITY_EXT + RATE_EXT == WIDTH_EXT, "capacity+rate mismatch");
            assert!(WIDTH_EXT * D == WIDTH, "width extension mismatch");
        };
```

Reply:
nit, but just mentioning `const` is enough:


```rust
const {
    assert!(CAPACITY_EXT + RATE_EXT == WIDTH_EXT, "capacity+rate mismatch");
    assert!(WIDTH_EXT * D == WIDTH, "width extension mismatch");
}
```